src/core/boxes.c

```diff
@@ -1122,15 +1123,10 @@ meta_rectangle_find_linepoint_closest_to_point (double x1,
    * but that's all the below few lines do.
    */
 
-  double diffx, diffy, den;
-  diffx = x2 - x1;
-  diffy = y2 - y1;
-  den = diffx * diffx + diffy * diffy;
-
-  *valx = (py * diffx * diffy + px * diffx * diffx +
-           y2 * x1 * diffy - y1 * x2 * diffy) / den;
-  *valy = (px * diffx * diffy + py * diffy * diffy +
-           x2 * y1 * diffx - x1 * y2 * diffx) / den;
+   *valx = (px*pow(x1,2) - 2*px*x1*x2 + px*pow(x2,2) + x2*pow(y1,2) + x1*pow(y2,2) + (py*x1 - py*x2)*y1 - (py*x1 - py*x2 + (x1 + x2)*y1)*y2) /
+           (pow(x1,2) - 2*x1*x2 + pow(x2,2) + pow(y1,2) - 2*y1*y2 + pow(y2,2));
+   *valy = (py*pow(y1,2) + py*pow(y2,2) + (px*x1 - (px + x1)*x2 + pow(x2,2))*y1 - (px*x1 - pow(x1,2) - (px - x1)*x2 + 2*py*y1)*y2) /
+           (pow(x1,2) - 2*x1*x2 + pow(x2,2) + pow(y1,2) - 2*y1*y2 + pow(y2,2));
```

Comment:
This is unreadable...

Reply:
I can't decompose it into factors or temporary variables without adding a precision loss with floating-point numbers, like previous code did, probably by forcing base changes o whatever.

Reply:
Moreover, I don't know where the round-off error described in comment comes from, as it doesn't make reference to float.h

Reply:
In the end result is used as int - https://github.com/mate-desktop/marco/blob/38cd6190c8b2c1553fa814b23734439e0e95f577/src/core/constraints.c#L1343-L1350.

So I doubt precision loss is problem here...

Reply:
Yes, It performs truncation, not rounding, and it should be fixed, for instance 0.9999.... = 0.
I will do another PR fixing that.

Reply:
Use rounded values instead of truncate them https://github.com/mate-desktop/marco/pull/476